src/packages/dynamic-root/modals/dynamic-root-origin-picker-modal.element.ts

```diff
@@ -0,0 +1,110 @@
+import { UmbDocumentPickerContext } from '../../documents/documents/components/input-document/input-document.context.js';
+import { UmbModalBaseElement } from '@umbraco-cms/backoffice/modal';
+import { css, html, customElement, map, state, ifDefined } from '@umbraco-cms/backoffice/external/lit';
+import { umbExtensionsRegistry } from '@umbraco-cms/backoffice/extension-registry';
+import type { ManifestDynamicRootOrigin } from '@umbraco-cms/backoffice/extension-registry';
+import type { UmbTreePickerDynamicRoot } from '@umbraco-cms/backoffice/components';
+
+@customElement('umb-dynamic-root-origin-picker-modal')
+export class UmbDynamicRootOriginPickerModalModalElement extends UmbModalBaseElement {
+	@state()
+	private _origins: Array<ManifestDynamicRootOrigin> = [];
+
+	#documentPickerContext = new UmbDocumentPickerContext(this);
+
+	constructor() {
+		super();
+
+		this.#documentPickerContext.max = 1;
+
+		this.observe(umbExtensionsRegistry.byType('dynamicRootOrigin'), (origins: Array<ManifestDynamicRootOrigin>) => {
+			this._origins = origins;
+		});
+	}
+
+	#choose(item: ManifestDynamicRootOrigin) {
+		switch (item.meta.originAlias) {
+			// NOTE: Edge-case. Currently this is the only one that uses a document picker,
+			// but other custom origins may want other configuration options. [LK:2024-01-25]
+			case 'ByKey':
+				this.#openDocumentPicker(item.meta.originAlias);
+				break;
+			default:
+				this.#submit({ originAlias: item.meta.originAlias });
+				break;
+		}
+	}
+
+	#close() {
+		this.modalContext?.reject();
+	}
+
+	#openDocumentPicker(originAlias: string) {
+		this.#documentPickerContext
+			.openPicker({
+				hideTreeRoot: true,
+			})
+			.then(() => {
+				const selectedItems = this.#documentPickerContext.getSelection();
+				if (selectedItems.length !== 1) return;
+				this.#submit({
+					originAlias,
+					originKey: selectedItems[0],
+				});
+			});
+	}
```

Comment:
Consider implementing an async/await pattern here, which should ease the nested complexity a bit

Reply:
@iOvergaard I wasn't sure how to tackle this. Do we have any similar examples in the codebase?

Reply:
Haven't tested it but does something like this work?

```suggestion
	async #openDocumentPicker(originAlias: string): Promise<void> {
		await this.#documentPickerContext
			.openPicker({
				hideTreeRoot: true,
			});
		const selectedItems = this.#documentPickerContext.getSelection();
		if (selectedItems.length !== 1) return;
		this.#submit({
			originAlias,
			originKey: selectedItems[0],
		});
	}
```

Reply:
@iOvergaard Would it be something like this...?

```
async #openDocumentPicker(originAlias: string) {
	await this.#documentPickerContext.openPicker({ hideTreeRoot: true });

	const selectedItems = this.#documentPickerContext.getSelection();
	if (selectedItems.length !== 1) return;

	this.#submit({
		originAlias,
		originKey: selectedItems[0],
	});
}
```

Reply:
Super, thanks!

Reply:
Implemented! 👍